### DISCUSSION TITLE
How to simulate multiple drive cycles? (using the modulo function)

### DISCUSSION THREAD

**Discussion by anandmy** [pybamm-team/PyBaMM#1293] [Q&A] [answered]
I am trying to simulate a drive cycle for a long time duration (at least 1 year). Currently I have a drive cycle representing a single day's current profile and want to replicate it for multiple days (years hopefully). While this is easier to imagine when using the Experiment class as it allows using lists, is there a way of doing the same for the Current Function [A] parameter or using a different approach?

Tino suggested using the modulo function (t % period) but I could not figure out how to use that. Any help would be appreciated. Thank you.

**Comment by brosaplanella** [comment]
You first need to define the interpolant from your dataset. You can find an example of how to do so in the "Loading current data section" of the notebook about [how to change the input current](https://github.com/pybamm-team/PyBaMM/blob/develop/examples/notebooks/change-input-current.ipynb).

Note that when you create the interpolant
```python3
current_interpolant = pybamm.Interpolant(drive_cycle, timescale * pybamm.t)
```
you pass as the variable the dimensional time `timescale * pybamm.t`. In your case, you want to take the dimensional time [modulo](https://en.wikipedia.org/wiki/Modulo_operation) the period. In your case, the period is the duration of the drive cycle, so you can probably read that from the data as well. Assuming that you have already defined the `period` you can then write
```python3
current_interpolant = pybamm.Interpolant(drive_cycle, (timescale * pybamm.t) % period)
```

Note that you can call the modulo operator both by using `%` and `pybamm.Modulo(left, right)`. See [the documentation](https://pybamm.readthedocs.io/en/latest/source/expression_tree/binary_operator.html?highlight=modulo#pybamm.Modulo) for more information.

**Comment by valentinsulzer** [comment, marked as answer]
The following example repeats the US06 drive cycle 3 times (for 1800s instead of 600s):
```python3
import pybamm
import pandas as pd
import os

os.chdir(pybamm.__path__[0] + "/..")

pybamm.set_logging_level("INFO")

# load model and update parameters so the input current is the US06 drive cycle
model = pybamm.lithium_ion.SPMe({"thermal": "lumped"})
param = model.default_parameter_values


# import drive cycle from file
drive_cycle = pd.read_csv(
    "pybamm/input/drive_cycles/US06.csv", comment="#", header=None
).to_numpy()

# create interpolant
timescale = param.evaluate(model.timescale)
# in the following line we use % 600 to repeat the drive cycle every 600 seconds
current_interpolant = pybamm.Interpolant(drive_cycle, (timescale * pybamm.t) % 600)

# set drive cycle
param["Current function [A]"] = current_interpolant


# create and run simulation using the CasadiSolver in "fast" mode, remembering to
# pass in the updated parameters
sim = pybamm.Simulation(
    model, parameter_values=param, solver=pybamm.CasadiSolver(mode="fast")
)
# we need to specify the solve time to be different from the default [0, 600]
# this will raise a warning but this can be ignored
sim.solve([0, 1800])
# use this instead to get rid of the warning
# t_drive = drive_cycle[:, 0]
# step_size = np.min(np.diff(t_drive))
# t_eval = np.arange(0, 1800, step_size)
# sim.solve(t_eval)
# you can play with step_size to change the level of resolution of the drive cycle

sim.plot(
    [
        "Negative particle surface concentration [mol.m-3]",
        "Electrolyte concentration [mol.m-3]",
        "Positive particle surface concentration [mol.m-3]",
        "Current [A]",
        "Negative electrode potential [V]",
        "Electrolyte potential [V]",
        "Positive electrode potential [V]",
        "Terminal voltage [V]",
        "X-averaged cell temperature",
    ]
)
```

**Reply by ssingh-ipa** [reply]
While running the model, one driving cycle can be repeated by multiplying the driving cycles through % xx. However, is it possible to run a model for two different cycles consecutively? 

For e.g. DrivingCycle1(DC1) = US06.csv and DrivingCycle2(DC2) = GridStorage.csv 
Firstly I execute the model for [0,3600] with DC1 and then I want to start the next simulation from 3601 with DC2 and with model variables that were at the end of the first simulation. I am trying to do this using options such as update() and new_copy() but they have not worked so far. What would you suggest?

**Reply by valentinsulzer** [reply]
Just make a new csv file that has the drive cycles one after the other?

**Reply by ssingh-ipa** [reply]
Actually this is not just about 2 driving cycle but rather large sets of driving data and operating conditions over long periods. This is an effort of updating the model along with the battery.  Is there a command to start the simulation from that time and state when the previous simulation stopped?

**Reply by ssingh-ipa** [reply]
Maybe this will help with some context: 
model = pybamm.lithium_ion.DFN()
sim1 = pybamm.Simulation(model, parameter_values=parameter_values)
#..Executed for DC1 in sim1
model_upd._solution = sim1.solution.last_state

Would this retain the previous model state? 

I cannot find many examples of using the attributes explained in the following documentations.  https://pybamm.readthedocs.io/en/latest/source/solvers/solution.html https://pybamm.readthedocs.io/en/latest/source/solvers/processed_variable.html

**Reply by valentinsulzer** [reply]
`sim.solve(starting_solution=sol)`
https://github.com/pybamm-team/PyBaMM/blob/develop/examples/notebooks/simulating-long-experiments.ipynb

**Reply by valentinsulzer** [reply]
Or
```python3
t = timescale * pybamm.t
current_interpolant = (t <= 100) * pybamm.Interpolant(drive_cycle_1, t) + (t > 100) * pybamm.Interpolant(drive_cycle_2, t)
```

**Reply by valentinsulzer** [reply]
using code to concatenate csv files is also pretty easy

**Reply by ssingh-ipa** [reply]
> `sim.solve(starting_solution=sol)`
> https://github.com/pybamm-team/PyBaMM/blob/develop/examples/notebooks/simulating-long-experiments.ipynb

Looks like that starting_solution is only valid for experimental inputs and not for driving cycles. 

Also, executing https://github.com/pybamm-team/PyBaMM/blob/develop/examples/notebooks/simulating-long-experiments.ipynb currently gives me the following error - 
"esoh_model = pybamm.lithium_ion.ElectrodeSOH()
AttributeError: module 'pybamm.models.full_battery_models.lithium_ion' has no attribute 'ElectrodeSOH'"

**Reply by valentinsulzer** [reply]
>Also, executing https://github.com/pybamm-team/PyBaMM/blob/develop/examples/notebooks/simulating-long-experiments.ipynb currently gives me the following error -
"esoh_model = pybamm.lithium_ion.ElectrodeSOH()
AttributeError: module 'pybamm.models.full_battery_models.lithium_ion' has no attribute 'ElectrodeSOH'"

make sure you are using the develop branch of pybamm if testing examples on the develop branch. If you installed with pip look for examples on the main branch

**Reply by asifsultan-THD** [reply]
I want to simulate it for drive cycle but in the above lines I am facing issue

timescale = param.evaluate(model.timescale)
current_interpolant = pybamm.Interpolant(drive_cycle, timescale * pybamm.t)

---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
C:\Users\AAA27~1.SUL\AppData\Local\Temp/ipykernel_2976/286842135.py in <module>
      2 timescale = param.evaluate(model.timescale)
      3 in the following line we use % 600 to repeat the drive cycle every 600 seconds
----> 4 current_interpolant = pybamm.Interpolant(drive_cycle, timescale * pybamm.t) % 600)

TypeError: __init__() missing 1 required positional argument: 'children'

**Reply by valentinsulzer** [reply]
In the latest versions, you have to do `Interpolant(data[0], data[1], timescale * pybamm.t)`

**Reply by asifsultan-THD** [reply]
Thanks for the reply. Actually I am stuck in one of your more line. Its getting simulated. But its not getting solved and giving runtime error. I am talking about this line:------         sim.solve([0, 1800])
Using the code below this line also giving the same error

And the error is :- RuntimeError: .../casadi/core/function_internal.cpp:145: Error calling BSplineInterpolant::init for 'LUT':
.../casadi/solvers/bspline_interpolant.cpp:91: Assertion "x.size()>=2*m+2" failed: Need more data points

Please correct it just been fews days on Pybamm
Thanks in Advance

**Comment by Gokarna-Bahadur-Thapa** [comment]
Help me to correct this code
# create interpolant
timescale = param.evaluate(model.timescale)
# in the following line we use % 600 to repeat the drive cycle every 600 seconds
current_interpolant = pybamm.Interpolant(drive_cycle, (timescale * pybamm.t) % 600)

NotImplementedError                       Traceback (most recent call last)
Cell In[10], line 2
      1 # create interpolant
----> 2 timescale = param.evaluate(model.timescale)
      3 # in the following line we use % 600 to repeat the drive cycle every 600 seconds
      4 current_interpolant = pybamm.Interpolant(drive_cycle, (timescale * pybamm.t) % 600)

File ~\anaconda3\Lib\site-packages\pybamm\models\base_model.py:294, in BaseModel.timescale(self)
    292 @property
    293 def timescale(self):
--> 294     raise NotImplementedError(
    295         "timescale has been removed since models are now dimensional"
    296     )

NotImplementedError: timescale has been removed since models are now dimensional

**Reply by agriyakhetarpal** [reply]
Here is a related discussion about how to adapt your code around the removal of `timescale` due to dimensional models: https://github.com/pybamm-team/PyBaMM/discussions/3712
